### PR TITLE
[multistage] hybrid routing support

### PR DESCRIPTION
**Pull request by walterddr**
Overview
===
This PR is preliminary support for hybrid routing on multistage engine. It copies so of the logics in the broker request side but not all. and only has basic testing.

Details
===
- copied some time boundary logics (manager and associated logics) into either core module or replicated in multi-stage planner
- created a hybrid routing dispatchable stagemetadata 
  - indexed each segment with table types
  - added time boundary info if dispatching multiple table types, otherwise ignore
  - reconstructed server requests multiple times. 

TODOs
===
this duplicates lots of logics out of BaseBrokerRequestHandler and also it doesn't do pruning well, so follow up
- consolidate the logic and abstract them out as util from BaseBrokerRequestHandler
- create smarter pruning logic on multi-stage server/leaf request compiler.

### CODE REVIEW

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/QueryRunner.java [hunk 1]
```diff
@@ -19,8 +19,10 @@
 package org.apache.pinot.query.runtime;
 
 import java.nio.ByteBuffer;
+import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/QueryRunner.java:25]:
```suggestion
```

File: pinot-common/src/main/proto/worker.proto [hunk 2]
```diff
@@ -68,7 +68,13 @@ message StagePlan {
 message StageMetadata {
   repeated string instances = 1;
   repeated string dataSources = 2;
-  map<string, SegmentList> instanceToSegmentList = 3;
+  map<string, SegmentMap> instanceToSegmentMap = 3;
```
Comment (kishoreg) [inline pinot-common/src/main/proto/worker.proto:71]: we can add another map<string,SegmentList> right? one for real time and one for offline.. thought your approah is more flexible if we add more tabletype

File: pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java [hunk 3]
```diff
@@ -58,11 +60,37 @@ public WorkerManager(String hostName, int port, RoutingManager routingManager) {
 
   public void assignWorkerToStage(int stageId, StageMetadata stageMetadata) {
     List<String> scannedTables = stageMetadata.getScannedTables();
-    if (scannedTables.size() == 1) { // table scan stage, need to attach server as well as segment info.
-      RoutingTable routingTable = getRoutingTable(scannedTables.get(0));
-      Map<ServerInstance, List<String>> serverInstanceToSegmentsMap = routingTable.getServerInstanceToSegmentsMap();
+    if (scannedTables.size() == 1) {
+      // table scan stage, need to attach server as well as segment info for each physical table type.
+      Map<String, RoutingTable> routingTableMap = getRoutingTable(scannedTables.get(0));
```
Comment (kishoreg) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java:65]: is scannedtable referring to logical table name or tablename with type.. can we name variable to indicate that?

File: pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java [hunk 4]
```diff
@@ -86,13 +114,23 @@ private static List<ServerInstance> filterServers(Collection<ServerInstance> ser
     return serverInstances;
   }
 
-  private RoutingTable getRoutingTable(String tableName) {
+  private Map<String, RoutingTable> getRoutingTable(String tableName) {
```
Comment (kishoreg) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java:117]: tableName -> tableNameWithType right?

**Review by kishoreg — APPROVED** [review]

File: pinot-common/src/main/proto/worker.proto (hunk 2 above)
Comment (walterddr) [reply pinot-common/src/main/proto/worker.proto:71]: yes this way it is more flexible

**Pull request merged**